src/plugins/shared_dictionaries_plugin/shared_dictionaries_column_processor.hpp

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include "hyrise.hpp"
+#include "shared_dictionaries_plugin.hpp"
+#include "storage/table.hpp"
+#include "utils/size_estimation_utils.hpp"
+
+namespace opossum {
+
+/**
+ * Finds and merges dictionaries with shared data for the given column of a table.
+ * Each dictionary segment is compared based on its similarity with all other existing shared dictionaries.
+ * If the similarity (e.g. jaccard-index) is above the threshold it merges the dictionaries.
+ * If none of the existing shared dictionaries meet the threshold it compares the current dictionary with the
+ * previous dictionary and eventually merges them to create a new shared dictionary.
+ */
+template <typename T>
+class SharedDictionariesColumnProcessor {
+  friend class SharedDictionariesPluginTest;
+
+ public:
+  /**
+ * Merge plan for a shared dictionary containing the segments to merge and additional information
+ */
+  class MergePlan {
+   public:
+    std::shared_ptr<const pmr_vector<T>> shared_dictionary;
+    std::vector<SegmentChunkPair<T>> segment_chunk_pairs_to_merge = {};
+    bool contains_non_merged_segment = false;
+    bool contains_already_merged_segment = false;
+    uint64_t non_merged_total_bytes = 0;
+    uint64_t non_merged_dictionary_bytes = 0;
+
+    explicit MergePlan(const std::shared_ptr<const pmr_vector<T>>& init_shared_dictionary)
+        : shared_dictionary(init_shared_dictionary) {}
+
+    void add_segment_chunk_pair(const SegmentChunkPair<T>& segment_chunk_pair, bool is_already_merged);
+  };
+
+  const std::shared_ptr<Table> table;
+  const std::string table_name;
+  const ColumnID column_id;
+  const std::string column_name;
+  const float jaccard_index_threshold;
+  SharedDictionariesPlugin::SharedDictionariesStats& stats;
+
+  SharedDictionariesColumnProcessor(const std::shared_ptr<Table>& init_table, const std::string& init_table_name,
+                                    const ColumnID init_column_id, const std::string& init_column_name,
+                                    const float init_jaccard_index_threshold,
+                                    SharedDictionariesPlugin::SharedDictionariesStats& init_stats);
```

Comment:
If the Plugin is unloaded while the processor is working, the accesses to stats could access memory that was already freed, right?